include/onnxruntime/core/framework/execution_provider.h

```diff
@@ -299,6 +300,16 @@ class IExecutionProvider {
   /** Does the EP support concurrent calls to InferenceSession::Run to execute the model.
    */
   virtual bool ConcurrentRunSupported() const { return true; }
+  /** Some session option values (default or user provided) may not work with some EPs.
+   * The EP should alter the SessionOptions object fit its implementation to avoid potential crash.
+   * Rather than put the onus on the user to know these, make the appropriate change while logging the change.
+   *
+   * @param session_options The SessionOptions object to be altered.
+   * @param logger The EP write the reason of the altering to it.
+   */
+  virtual common::Status CheckSessionOptions(const SessionOptions& /*session_options*/) {
```

Comment:
nit: add an empty line between the previous function and the start of the comment for this one.

The comment needs updating as well as the EP should not alter the SessionOptions object.
I would suggest adding an `@return` saying the EP should return an error if it's incompatible with the SessionOptions that specifies which option is problematic so the user knows what needs to be fixed.

'Check' doesn't really convey that the EP might update its internals based on the session options, which is why this method is non-const. Maybe ProcessSessionOptions would be slightly better.

Reply:
We do have process EP itself in this function,  so we can't mark this method as const.
This function should works with two purpose: 1 check if EP could work SessionOptions, 2. Adjust EP-self to make compatiable to run normally.  The first one is  compulsory, the second is not. How about combine them together? say `CheckSessionOptionsAndProcess`

Reply:
'Process' is fairly generic. 'checking' is a form of processing, as is updating say threadpool sizes. I don't think anything is 'compulsory' - it's up to the EP to decide what it wants to do, so I'm not sure I see value in the longer name. If it was compulsory I would have expected every single EP was required to implement it.

Reply:
Can we rename it to ValidateSessionOptions?

Reply:
Done.

Reply:
I guess we don't need this API any more since we agreed to go with the factory based solution, right?

Reply:
Not quite sure if this condition based on EP-in-factory  solution works as expected.

For example, we can create factory and EP with python-binding
https://github.com/microsoft/onnxruntime/blob/aefc82b5c99671b527ff0196cce9c5d002d71343/onnxruntime/python/onnxruntime_pybind_schema.cc#L65
However, it didn't touch any sessionOptions here, which means we have to pass nullptr to `xxFactory::create`, Then we create EP successfully, But it will lead to troubles or even crash.

Any thoughs on this question? Thanks